src/Pos/Communication/Limits.hs

```diff
@@ -0,0 +1,455 @@
+
+instance Limiter (Limit t) where
+    limitGet (Limit l) = Bi.limitGet $ fromIntegral l
+    addLimit a = mappend (Limit a)
+
+-- | Bounds `InvOrData`.
+instance Limiter l => Limiter (Limit t, l) where
+    limitGet (invLimit, dataLimits) parser = do
+        lookAhead getWord8 >>= \case
+            0   -> limitGet invLimit parser
+            1   -> limitGet dataLimits parser
+            tag -> fail ("get@InvOrData: invalid tag: " ++ show tag)
+
+    addLimit a (l1, l2) = (a `addLimit` l1, a `addLimit` l2)
+
+-- | Bounds `DataMsg`.
+-- Limit depends on value of first byte, which should be in range @0..3@.
+instance Limiter (Limit t, Limit t, Limit t, Limit t) where
+    limitGet limits parser = do
+        tag <- fromIntegral <$> lookAhead getWord8
+        case (limits ^.. each) ^? ix tag of
+            Nothing -> fail ("get@DataMsg: invalid tag: " ++ show tag)
+            Just limit -> limitGet limit parser
+
+    addLimit a = both %~ addLimit a
+
+-- | Specifies limit on message length.
+-- Deserialization would fail if incoming data size exceeded this limit.
+-- At serialisation stage message size is __not__ checked.
+class Limiter (LimitType a) => MessageLimited a where
+    type LimitType a :: *
+    getMsgLenLimit :: DB.MonadDBLimits m => Proxy a -> m (LimitType a)
+
+instance MessageLimited (MsgBlock ssc) where
+    type LimitType (MsgBlock ssc) = Limit (MsgBlock ssc)
+    getMsgLenLimit _ = Limit <$> DB.getMaxBlockSize
+
+instance MessageLimited MsgGetBlocks where
+    type LimitType MsgGetBlocks = Limit MsgGetBlocks
+    getMsgLenLimit _ = return msgLenLimit
+
+instance MessageLimited MsgGetHeaders where
+    type LimitType MsgGetHeaders = Limit MsgGetHeaders
+    getMsgLenLimit _ = return $
+        MsgGetHeaders <$> vector Const.genesisMaxGetHeadersNum <*> msgLenLimit
+
+instance MessageLimited (MsgHeaders ssc) where
+    type LimitType (MsgHeaders ssc) = Limit (MsgHeaders ssc)
+    getMsgLenLimit _ = return $
+        MsgHeaders <$> vectorOf Const.blkSecurityParam
+            (Limit Const.genesisMaxHeaderSize)
+
+instance MessageLimited (InvMsg key tag) where
+    type LimitType (InvMsg key tag) = Limit (InvMsg key tag)
+    getMsgLenLimit _ = return msgLenLimit
+
+instance MessageLimited (ReqMsg key tag) where
+    type LimitType (ReqMsg key tag) = Limit (ReqMsg key tag)
+    getMsgLenLimit _ = return msgLenLimit
+
+instance MessageLimited (DataMsg TxMsgContents) where
+    type LimitType (DataMsg TxMsgContents) = Limit (DataMsg TxMsgContents)
+    getMsgLenLimit _ = return msgLenLimit
+
+instance MessageLimited (DataMsg UpdateVote) where
+    type LimitType (DataMsg UpdateVote) = Limit (DataMsg UpdateVote)
+    getMsgLenLimit _ = return msgLenLimit
+
+instance MessageLimited (DataMsg (UpdateProposal, [UpdateVote])) where
+    type LimitType (DataMsg (UpdateProposal, [UpdateVote])) =
+        Limit (DataMsg (UpdateProposal, [UpdateVote]))
+    getMsgLenLimit _ = return msgLenLimit
+
+noLimit :: Limit a
+noLimit = 1000000
```

Comment:
> Is it value such that we can easily parse so many bytes and at the same time it can fit all fixed-size data? 
Yes. But ok, I see now this is bad constant, I'll remove it